geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/Z3IdxStrategy.scala

```diff
@@ -126,57 +126,51 @@ class Z3IdxStrategy(val filter: QueryFilter) extends Strategy with Logging with
     val env = geometryToCover.getEnvelopeInternal
     val (lx, ly, ux, uy) = (env.getMinX, env.getMinY, env.getMaxX, env.getMaxY)
 
-    def qp(week: Int, lt: Long, ut: Long, contained: Boolean) =
-      queryPlanForPrefix(week, (lx, ux), (ly, uy), (lt, ut), z3table,
-        kvsToFeatures, iterators, colFamily, numThreads, contained)
-
     val epochWeekStart = Weeks.weeksBetween(Z3Table.EPOCH, interval.getStart)
     val epochWeekEnd = Weeks.weeksBetween(Z3Table.EPOCH, interval.getEnd)
     val weeks = scala.Range.inclusive(epochWeekStart.getWeeks, epochWeekEnd.getWeeks)
     val lt = Z3Table.secondsInCurrentWeek(interval.getStart, epochWeekStart)
     val ut = Z3Table.secondsInCurrentWeek(interval.getEnd, epochWeekEnd)
 
-    // the z3 index breaks time into 1 week chunks, so create a query plan for each week in our range
-    if (weeks.length == 1) {
-      Seq(qp(weeks.head, lt, ut, contained = false))
+    val lz = Z3_CURVE.index(lx, ly, lt).z
+    val uz = Z3_CURVE.index(ux, uy, ut).z
+
+    // the z3 index breaks time into 1 week chunks, so create a range for each week in our range
+    val (ranges, zMap) = if (weeks.length == 1) {
+      val ranges = getRanges(weeks, (lx, ux), (ly, uy), (lt, ut))
+      val map = Map(weeks.head.toShort -> (lz, uz))
+      (ranges, map)
     } else {
-      val head +: xs :+ last = weeks.toList
       // time range for a chunk is 0 to 1 week (in seconds)
-      val timeMax = Weeks.ONE.toStandardSeconds.getSeconds
-      val startQP = qp(head, lt, timeMax, contained = false)
-      val endQP = qp(last, 0, ut, contained = false)
-      val middleQPs = xs.map(w => qp(w, 0, timeMax, contained = true))
-      Seq(startQP, endQP) ++ middleQPs
+      val tMax = Weeks.ONE.toStandardSeconds.getSeconds
+      val head +: middle :+ last = weeks.toList
+      val headRanges = getRanges(Seq(head), (lx, ux), (ly, uy), (lt, tMax))
+      val lastRanges = getRanges(Seq(last), (lx, ux), (ly, uy), (0, ut))
```

Comment:
last will always not be empty due to the length check just above - middle might be empty and we check for that.
scala> val head +: xs :+ last = List(1, 2)
head: Int = 1
xs: List[Int] = List()
last: Int = 2